### DISCUSSION TITLE
How do I manage sensitive information?

### DISCUSSION THREAD

**Discussion by caffeinism** [argoproj/argo-cd#5133] [Q&A] [answered]
Usually sensitive information (password, credential, ...) is not recommended to be managed with git.

How to deploy safely and conveniently in this case?

**Comment by caffeinism** [comment]
I have found this manual.
https://argoproj.github.io/argo-cd/operator-manual/secret-management/
Sorry for the duplicate question.

**Reply by sbose78** [reply]
Which one did you eventually go for? :)

**Reply by sbose78** [reply]
Note, typically there are two approaches
1. Encrypt secrets in git.
2. Reference secrets in git.

**Reply by caffeinism** [reply]
> Note, typically there are two approaches
> 
> 1. Encrypt secrets in git.
> 2. Reference secrets in git.

I chose option 2. 

The reason is that the secrets always can be changed. "Is it necessary to keep a record of changes to secrets in git?" When asked by myself, I thought not. Another reason was that it felt more desirable to manage secrets in a centralized place.

Of course it has the downside of being stateful. But I thought the advantage was bigger.

(But I'm new to this, I don't know well, and I'm doing it for practice. It's just my thoughts. 😛)

**Reply by sbose78** [reply]
Which tool/project do you use for referencing secrets?

**Reply by sbose78** [reply]
Where are your secrets being stored?

**Reply by caffeinism** [reply]
It was written like that, but I eventually turned to using SealedSecret. It felt much cleaner and idempotent in deployment management as well.

**Comment by greenkiwi** [comment, marked as answer]
We took approach 1.

We have secret templates that contain the secrets that we want to have present and a reference to that secret in our secret store.  When we run a CLI app that generates `SealedSecrets` and stores these in git.  We take a hash of the secrets and use that to name the secret and update appropriate patches.

To this end, we have a process that is very similar to the way the configmap generator works.  Any time a secret changes, we have to perform a rollout to get the new values.   We liked it since it made it really clear when a change was going out and would allow for rolling that change back (hopefully) seamlessly.

We're looking at sharing our `kubesealer` CLI app.

**Reply by sbose78** [reply]
Thanks for sharing, how do you handle key rotation ?

**Reply by greenkiwi** [reply]
@sbose78 rotating kubesealer keys?  Or rotating secrets stored in kubesealer?

Just to provide some more context, our `SecretTemplates` look like this:
```
apiVersion: turo.com/V1
kind: SecretTemplate
metadata:
  name: newrelic-secrets
  namespace: some-namespace
spec:
  certName: cert-name
  disableNameSuffixHash: "true"|"false"
  data:
    NEW_RELIC_LICENSE_KEY:
      path: /newrelic/license_key
```

And that is used to generate the `SealedSecret`.

For sealing key renewal, the keys for each cluster are downloaded and made available to our command.  This means that they can be up to date with each sealing.  Note that we don't update the secrets every time we process them.  This is so that we can see when individual secrets change.  That would be pretty easy to change/update.

For user secrets, we always have both the username/key id and then the secret in our secret store.  Sometimes they are just updated in place, other times we add a `_v0` to the two values.  In both cases, we process all our templates and would notice that the values had changed and a new secret was available and we would rollout the changes with an updated secret name.
